src/core/symbology/qgsmarkersymbollayer.cpp

```diff
@@ -2683,24 +2683,47 @@ void QgsRasterMarkerSymbolLayer::renderPoint( QPointF point, QgsSymbolRenderCont
   if ( !p )
     return;
 
-  bool hasDataDefinedSize = false;
-  double scaledSize = calculateSize( context, hasDataDefinedSize );
-  double width = context.renderContext().convertToPainterUnits( scaledSize, mSizeUnit, mSizeMapUnitScale );
-  bool hasDataDefinedAspectRatio = false;
-  double aspectRatio = calculateAspectRatio( context, scaledSize, hasDataDefinedAspectRatio );
-  double height = width * ( preservedAspectRatio() ? defaultAspectRatio() : aspectRatio );
-
-  //don't render symbols with size below one or above 10,000 pixels
-  if ( static_cast< int >( width ) < 1 || 10000.0 < width )
-  {
-    return;
-  }
-
   QString path = mPath;
   if ( mDataDefinedProperties.isActive( QgsSymbolLayer::PropertyName ) )
   {
     context.setOriginalValueVariable( mPath );
     path = mDataDefinedProperties.valueAsString( QgsSymbolLayer::PropertyName, context.renderContext().expressionContext(), mPath );
+  }
+
+  if ( path.isEmpty() )
+    return;
+
+  double width = 0.0;
+  double height = 0.0;
+  double scaledSize = 0.0;
+  double aspectRatio = 0.0;
+
+  // RenderPercentage Unit Type takes original image size
+  if ( mSizeUnit == QgsUnitTypes::RenderPercentage )
+  {
+    QSize size = QgsApplication::imageCache()->originalSize( path );
+    if ( size.isNull() || !size.isValid() || size.width() == 0 || size.height() == 0 )
+      return;
+
+    scaledSize = ( mSize * static_cast<double>( size.width() ) ) / 100.0;
+    width = context.renderContext().convertToPainterUnits( scaledSize, mSizeUnit, mSizeMapUnitScale );
+
+    height = ( mSize * static_cast<double>( size.height() ) ) / 100.0;
+    height = context.renderContext().convertToPainterUnits( height, mSizeUnit, mSizeMapUnitScale );
+
+    // don't render symbols with size below one or above 10,000 pixels
+    if ( static_cast< int >( width ) < 1 || 10000.0 < width || static_cast< int >( height ) < 1 || 10000.0 < height )
+      return;
+  }
+  else
+  {
+    bool hasDataDefinedSize = false;
+    scaledSize = calculateSize( context, hasDataDefinedSize );
```

Comment:
You'll want to move this up to before you check for percentage based sizes, otherwise data defined sizes in percent won't work